File: briareos-master/dev/project/briareos/core/worker_manager.py

from briareos.utils import logger
from briareos.core import distributed_zmq

from threading import Thread

import zmq
import docker
import time
import uuid

USAGE_MSG = "REPORT USAGE"
START_NEW_INSTANCE_MSG = "NEW INSTANCE"
STOP_INSTANCE_MSG = "STOP INSTANCE"

# TODO containers: limit CPU and RAM
# TODO thread lock
# TODO choose a cluster to start new instance


class WorkerManager:
    metric_interval = 15  #TODO config
    lower_bound = 10
    upper_bound = 70

    def __init__(self, zbroker, worker_manager_config):
        self._zbroker = zbroker
        self._context = zmq.Context()
        self._init_sockets(worker_manager_config)
        self.interval = int(worker_manager_config.get("interval"))  # TODO config and default value
        self.worker_stats = {}
        self.connected_zworkers = []  # TODO connected_*z*workers (distributed_zmq.py)
        self.average_cpu = None
        self.average_memory = None
        self.already_requested = False
        self.n_workers = 0
        self.clusters = {}

    def _init_sockets(self, worker_manager_config):
        ip = worker_manager_config.get("ip")
        port = worker_manager_config.get("port")
        sink_port = worker_manager_config.get("sink_port")

        if ip is None:
            ip = "*"

        self.publisher_address = distributed_zmq.get_address(ip, port)
        self.sink_address = distributed_zmq.get_address(ip, sink_port)
        self._publisher = self._context.socket(zmq.PUB)
        self._sink = self._context.socket(zmq.PULL)

    def start(self):
        logger.subinfo("Worker Manager address: '%s' | '%s'" % (self.publisher_address, self.sink_address))
        self._publisher.bind(self.publisher_address)
        self._sink.bind(self.sink_address)

        run_publisher_thread = Thread(target=self.run_publisher)
        run_publisher_thread.daemon = True
        run_publisher_thread.start()

        run_sink_thread = Thread(target=self.run_sink)
        run_sink_thread.daemon = True
        run_sink_thread.start()

        run_thread = Thread(target=self.run)
        run_thread.daemon = True
        run_thread.start()

    # TODO stop

    def run_publisher(self):
        while True:
            self._publisher.send(USAGE_MSG)
            time.sleep(self.interval+0.01)
            # self._remove_disconnected_workers()

    def run_sink(self):
        while True:
            cluster_id = self._sink.recv()
            stats = self._sink.recv_json()
            self._process_usage_stats(cluster_id, stats)


    def run(self):
        while True:
            # logger.debug(self.connected_zworkers)
            try:
                self._sliding_windows()
            except:
                pass

            if self.average_cpu is not None and self.average_memory is not None:
                logger.debug("Metrics: avg_cpu=%s avg_memory=%s" % (self.average_cpu, self.average_memory))
                if self.average_cpu >= self.upper_bound or self.average_memory >= self.upper_bound:
                    if not self.already_requested:
                        print("Trying to start worker")
                        
                        cluster_id = self.choose_cluster_id()
                        self.clusters[cluster_id] += 1
                        self._publisher.send(START_NEW_INSTANCE_MSG + " " + cluster_id)
                        self.already_requested = True
                        

                    # self._publisher.send(cluster_id)
                elif self.average_cpu <= self.lower_bound or self.average_memory <= self.lower_bound:
                    if self.n_workers > 1:
                        '''
                        pass
                        print("Trying to stop worker")
                        self._publisher.send(STOP_INSTANCE_MSG)
                        self.n_workers -= 1
                        #self._publisher.send(cluster_id)
                        '''

            time.sleep(self.interval)
            logger.debug("Tasks: %s | workers: %s" % (self._zbroker.broker.n_tasks, self.n_workers))

    def choose_cluster_id(self):
        '''
        clusters = {}
        for k in self.worker_stats:
            cluster_id = self.worker_stats[k][0]["cluster"]
            try:
                clusters[cluster_id] += 1
            except:
                clusters[cluster_id] = 1
        '''

        cluster_id = ""
        min_workers = 0xffffffff
        for c in self.clusters:
            if self.clusters[c] < min_workers:
                cluster_id = c
                min_workers = self.clusters[c]
        print(cluster_id),
        print(min_workers)
        return cluster_id


    def _sliding_windows(self):
        cpu_values = []
        memory_values = []
        worker_stats_to_remove = []

        current_time = time.time()
        for worker_id in self.worker_stats:
            stats = self.worker_stats[worker_id]
            for stat in stats:
                stat_time = stat["time"]
                if int(current_time - stat_time) <= self.metric_interval:
                    cpu_values.append(stat["cpu"])
                    memory_values.append(stat["memory"])
                else:
                    worker_stats_to_remove.append(worker_id)

        for worker_id in worker_stats_to_remove:
            del self.worker_stats[worker_id]

        self.average_cpu = average(cpu_values)
        self.average_memory = average(memory_values)

    def _process_usage_stats(self, cluster_id, stats):
        for zworker_id in stats:
            cpu_usage = stats[zworker_id]["cpu"]
            memory_usage = stats[zworker_id]["memory"]
            if zworker_id not in self.connected_zworkers:
                self.connected_zworkers.append(zworker_id)

            self._update_worker_stats(cluster_id, zworker_id, cpu_usage, memory_usage)

    def _update_worker_stats(self, cluster_id, zworker_id, cpu_usage, memory_usage):
        if self.worker_stats.get(zworker_id) is None:
            self.worker_stats[zworker_id] = []

        if self.clusters.get(cluster_id) is None:
            self.clusters[cluster_id] = 1

        self.worker_stats[zworker_id].append({"cluster": cluster_id, "cpu": cpu_usage,
                                              "memory": memory_usage, "time": time.time()})

    def _remove_disconnected_workers(self):
        worker_stats_to_remove = []
        for worker_id in self.worker_stats:
            if worker_id not in self.connected_zworkers:
                worker_stats_to_remove.append(worker_id)

        for worker_id in worker_stats_to_remove:
            del self.worker_stats[worker_id]

        self.connected_zworkers = []  # TODO sync or remove this...


class Cluster:
    docker_image_name = "zworker:latest"
    cpu_quota = 10000  # TODO config
    max_memory = 512  # TODO config

    def __init__(self, zreporter_config):
        self.id = uuid.uuid1()
        self._context = zmq.Context()
        self._init_sockets(zreporter_config)
        self._docker_client = docker.from_env()
        self.containers = []

    def _init_sockets(self, zreporter_config):
        worker_manager_config = zreporter_config.get("worker_manager")
        self.worker_manager_address = distributed_zmq.get_address(worker_manager_config.get("ip"), worker_manager_config.get("port"))
        self.sink_address = distributed_zmq.get_address(worker_manager_config.get("ip"), worker_manager_config.get("sink_port"))
        self._publisher = self._context.socket(zmq.SUB)
        self._publisher.setsockopt(zmq.SUBSCRIBE, "")
        self._sink = self._context.socket(zmq.PUSH)
        self.n_instances = 0

    def start(self):
	#n_workers = 2
	#for i in range(n_workers):
	#        self.start_new_instance()
        self.start_new_instance()

        logger.info("Connecting to Worker Manager: '%s' | '%s'" % (self.worker_manager_address, self.sink_address))
        self._publisher.connect(self.worker_manager_address)
        self._sink.connect(self.sink_address)
        run_thread = Thread(target=self.run)
        run_thread.daemon = True
        run_thread.start()

    def stop(self):
        self.stop_all_instances()

    def start_new_instance(self):
        logger.info("Starting new Z-Worker instance")
        self.n_instances += 1
        self.containers.append(self._docker_client.containers.run(Cluster.docker_image_name, detach=True,
                                                                  network_mode="host", cpu_quota=120000))
        # TODO network/config

    def stop_instance(self):
        if self.containers:
            logger.info("Stopping Z-Worker instance")
            self.containers[0].stop()
            self.containers.pop()

    def stop_all_instances(self):
        logger.info("Stopping all Z-Worker instances. Please wait...")
        for instance in self.containers:
            instance.stop()

    def run(self):
        while True:
            data = self._publisher.recv()
            print("Recv %s" % data)
            if data == USAGE_MSG:
                usage_stats = self.get_usage_stats()
                self._sink.send("%s" % self.id)
                self._sink.send_json(usage_stats)
            elif START_NEW_INSTANCE_MSG in data:
                cluster_id = data.split()[2]
                print(cluster_id),
                print(self.id)
                print(str(self.id) == cluster_id)
                if str(self.id) == cluster_id:
                    print("Ok %s" % self.n_instances)
                    if self.n_instances < 5:
                        logger.debug("Starting new Z-Worker instance")
                        self.start_new_instance()
            elif data == STOP_INSTANCE_MSG:
                logger.debug("Stopping one Z-Worker instance")
                self.stop_instance()

    def get_usage_stats(self):
        usage_stats = {}
        for container in self.containers:
            container_stats = container.stats(stream=False)
            container_id = container_stats["id"]
            memory_usage = self.get_memory_usage(container_stats)
            cpu_usage = self.get_cpu_usage(container_stats)
            usage_stats[container_id] = {"memory": memory_usage, "cpu": cpu_usage}
        return usage_stats

    @staticmethod
    def get_cpu_usage(container_stats):
        try:
            cpu_percent = 0.0
            precpu_stats = container_stats["precpu_stats"]
            cpu_stats = container_stats["cpu_stats"]
            cpu_total_usage = cpu_stats["cpu_usage"]["total_usage"]
            precpu_total_usage = precpu_stats["cpu_usage"]["total_usage"]
            system_cpu_usage = cpu_stats["system_cpu_usage"]
            presystem_cpu_usage = precpu_stats["system_cpu_usage"]
            percpu_usage = cpu_stats["cpu_usage"]["percpu_usage"]

            cpu_delta = cpu_total_usage - precpu_total_usage
            system_delta = system_cpu_usage - presystem_cpu_usage

            if cpu_delta > 0 and system_delta > 0:
                cpu_percent = (cpu_delta*1.0 / system_delta) * len(percpu_usage) * 100.0

            return cpu_percent
        except:
            return 0

    @staticmethod
    def get_memory_usage(container_stats):
        try:
            memory_stats = container_stats["memory_stats"]
            usage = memory_stats["usage"]
            limit = memory_stats["limit"]
            return usage*100.0/limit
        except:
            return 0


# TODO move this to utils?
def average(l):
    if not l:
        return None
    return sum(l) / float(len(l))

File: jbriareos-master/Tadeu_Freitas-jbriareos-c16b99c2c2d9/src/main/java/core/WorkerManager.java
package core;

import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;
import org.zeromq.SocketType;
import org.zeromq.ZContext;
import org.zeromq.ZMQ;
import utils.Loader;
import utils.WorkerStats;

import java.util.*;
import java.util.concurrent.Semaphore;

public class WorkerManager {
    ZContext context;
    ZMQ.Socket publisher;
    ZMQ.Socket sink;
    String publisherAddress;
    String sinkAddress;
    long interval;
    HashMap<String, LinkedList<WorkerStats>> workerStats;
    Double averageCPU;
    Double averageMem;
    public boolean alreadyRequested;
    public int numWorkers;
    public int numTasks;
    HashMap<UUID, Integer> workersPerCluster;

    private final Semaphore mutexPublisher = new Semaphore(1);

    public static final String USAGE_MSG = "REPORT USAGE";
    public static final String START_NEW_INSTANCE_MSG = "NEW INSTANCE";
    public static final String STOP_INSTANCE_MSG = "STOP INSTANCE";

    public static final int METRIC_INTERVAL = 15000;    // 15 secs
    public static final int LOWER_BOUND = 10;
    public static final int UPPER_BOUND = 70;

    Thread t1, t2, t3;

    public WorkerManager() {
        JSONObject brokerConfig = Loader.importZBrokerConfig();

        Object ipObj = brokerConfig.get("ip");
        String ip = "*";
        if (ipObj != null)
            ip = (String) ipObj;

        JSONObject config = (JSONObject) brokerConfig.get("worker_manager");
        long publisherPort = (Long) config.get("port");
        long sinkPort = (Long) config.get("sink_port");

        this.context = new ZContext();
        this.publisherAddress = DistributedZMQ.formAddress(ip, publisherPort);
        this.sinkAddress = DistributedZMQ.formAddress(ip, sinkPort);

        this.publisher = this.context.createSocket(SocketType.PUB);
        this.sink = this.context.createSocket(SocketType.PULL);

        this.interval = (long) config.get("interval");
        this.workerStats = new HashMap<>();
        this.averageCPU = null;
        this.averageMem = null;
        this.alreadyRequested = false;
        this.numWorkers = 0;
        this.numTasks = 0;
        this.workersPerCluster = new HashMap<>();
    }

    private Double average(LinkedList<Double> list) {
        if (list.isEmpty())
            return null;

        double sum = 0;
        for (double item : list)
            sum += item;

        return sum / list.size();
    }

    private void slidingWindows() {
        LinkedList<Double> cpuValues = new LinkedList<>();
        LinkedList<Double> memValues = new LinkedList<>();
        LinkedList<String> workerStatsToRemove = new LinkedList<>();

        long currentTime = new Date().getTime();
        for (Map.Entry<String, LinkedList<WorkerStats>> entry : this.workerStats.entrySet())
            for (WorkerStats stat : entry.getValue()) {
                if (currentTime - stat.revision <= METRIC_INTERVAL) {
                    cpuValues.addLast(stat.cpuUsage);
                    memValues.addLast(stat.memUsage);
                } else {
                    workerStatsToRemove.addLast(entry.getKey());
                }
            }

        for (String workerId : workerStatsToRemove)
            this.workerStats.remove(workerId);

        this.averageCPU = this.average(cpuValues);
        this.averageMem = this.average(memValues);
    }

    @SuppressWarnings("SameParameterValue")
    private UUID chooseClusterId(boolean withLeastWorkers) {
        String clusterId = "";
        int workers;

        if (withLeastWorkers)
            workers = 0xffffffff;
        else
            workers = 0;

        for (Map.Entry<UUID, Integer> entry : workersPerCluster.entrySet()) {
            if (entry.getValue() < workers) {
                clusterId = entry.getKey().toString();
                workers = entry.getValue();
            }
        }

        if (clusterId.isEmpty()) {
            System.out.println("There is no initialized cluster on the WorkerManager");
            System.exit(-1);
        }

        return UUID.fromString(clusterId);
    }

    private void updateWorkerStats(UUID clusterId, String zworkerContainerId, double cpuUsage, double memUsage) {
        WorkerStats stats = new WorkerStats(cpuUsage, memUsage);

        this.workerStats.computeIfAbsent(zworkerContainerId, k -> new LinkedList<>());
        this.workersPerCluster.putIfAbsent(clusterId, 1);

        LinkedList<WorkerStats> statsList = this.workerStats.get(zworkerContainerId);
        statsList.addLast(stats);
        this.workerStats.put(zworkerContainerId, statsList);
    }

    private void processUsageStats(UUID clusterId, JSONObject workerStats) {
        for (Object keyObj : workerStats.keySet()) {
            String containerId = (String) keyObj;

            JSONObject stats = (JSONObject) workerStats.get(containerId);
            double cpuUsage = (double) stats.get("cpu");
            double memUsage = (double) stats.get("memory");

            this.updateWorkerStats(clusterId, containerId, cpuUsage, memUsage);
        }
    }

    @SuppressWarnings("BusyWait")
    public void runPublisher() {
        while (!Thread.interrupted()) {
            try {
                mutexPublisher.acquire();

                this.publisher.send(USAGE_MSG);
                Thread.sleep(this.interval * 1000 + 10);
            } catch (RuntimeException | InterruptedException ignored) {
                return;
            } finally {
                mutexPublisher.release();
            }
        }
    }

    public void runSink() {
        JSONParser parser = new JSONParser();

        while (!Thread.interrupted()) {
            try {
                UUID clusterId = UUID.fromString(this.sink.recvStr());
                JSONObject workerStats = (JSONObject) parser.parse(this.sink.recvStr());
                this.processUsageStats(clusterId, workerStats);
            } catch (RuntimeException ignored) {
                return;
            } catch (ParseException ignored2) {
                System.out.println("Failure parsing worker statistics");
            }
        }
    }

    public void printMetrics() {
        System.out.println("\nMetrics:\n    AvgCPU = " + this.averageCPU + "\n    AvgMem = " + this.averageMem);
    }

    @SuppressWarnings("all")
    public void run() {
        while (!Thread.interrupted()) {
            this.slidingWindows();

            if (this.averageCPU != null && this.averageMem != null) {
                this.printMetrics();
                if (this.averageCPU >= UPPER_BOUND || this.averageMem >= UPPER_BOUND) {
                    if (!this.alreadyRequested) {
                        UUID clusterId = this.chooseClusterId(true);
                        int numWorkers = this.workersPerCluster.get(clusterId);
                        this.workersPerCluster.put(clusterId, numWorkers + 1);

                        try {
                            mutexPublisher.acquire();

                            this.publisher.send(START_NEW_INSTANCE_MSG + " " + clusterId);
                        } catch (RuntimeException | InterruptedException ignored) {
                            System.out.println("Couldn't send " + START_NEW_INSTANCE_MSG +
                                    " message to the cluster");
                            continue;
                        } finally {
                            mutexPublisher.release();
                        }

                        this.alreadyRequested = true;
                    }
                } else if ((this.averageCPU <= LOWER_BOUND && this.averageMem < UPPER_BOUND)
                        || this.averageMem <= LOWER_BOUND && averageCPU < UPPER_BOUND) {
                    /*
                        -> Acquire mutex and send STOP_INSTANCE_MSG to cluster
                        -> Send notice to broker about worker stopping, to remove it
                     */
                }
            }

            try {
                Thread.sleep(this.interval * 1000);
                System.out.println("Tasks: " + this.numTasks + " | Workers: " + this.numWorkers);
                System.out.println();
            } catch (InterruptedException ignored) {
                return;
            }
        }
    }

    public void start() {
        System.out.println("Worker Manager address: '" + this.publisherAddress + "' | '" + this.sinkAddress + "'");
        this.publisher.bind(this.publisherAddress);
        this.sink.bind(this.sinkAddress);

        t1 = new Thread(this::runPublisher);
        t2 = new Thread(this::runSink);
        t3 = new Thread(this::run);

        t1.start();
        t2.start();
        t3.start();
    }

    public void stop() {
        System.out.println("Stopping WorkerManager...");

        try {
            t1.interrupt();
            t2.interrupt();
            t3.interrupt();

            t1.join();
            t2.join();
            t3.join();
        } catch (Exception e) {
            e.printStackTrace();
        }

        this.publisher.close();
        this.sink.close();
        this.context.destroy();
    }
}
